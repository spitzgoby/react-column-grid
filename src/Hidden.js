import { addMissingSizes } from './breakpoints';
import React from 'react';
import { createUseStyles } from 'react-jss';
import styles from './Hidden.styles';

const defaultHide = false;

const Hidden = ({ children, hide }) => {
    const useShorthandSyntax = () => false;
    const adjustedHide = addMissingSizes('hide', hide, defaultHide, useShorthandSyntax);
    const useStyles = createUseStyles(styles);
    const { hidden: hiddenClass } = useStyles({ hide: adjustedHide });

    return (
        <span className={hiddenClass}>{children}</span>
    );
} 

export default Hidden;