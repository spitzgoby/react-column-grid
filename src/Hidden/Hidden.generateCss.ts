import { BreakpointsDefinition } from "../utils/breakpoints";
import { sizes } from "../utils/breakpoints";
import { generateScreenQueries } from "../utils/mediaQueries";

export const generateHiddenBreakpointCss = (breakpoints: BreakpointsDefinition) => {
    const sizeDefinitions = sizes.map((size) => `.rcg-h-${size}{display: none;}`);

    return generateScreenQueries(breakpoints, sizeDefinitions)
}