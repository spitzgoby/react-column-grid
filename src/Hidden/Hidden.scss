@use 'sass:map';
@import '../style/sizes.scss';

@media screen and (min-width: 1535px) {
  .rcg-h-xl {
    display: none;
  }
}

@each $sizeName, $minWidth in $minWidthSizes {
  @media screen and (min-width: #{$minWidth}) and (max-width: map.get($maxWidthSizes, $sizeName)) {
    .rcg-h-#{$sizeName} {
      display: none;
    }
  }
}