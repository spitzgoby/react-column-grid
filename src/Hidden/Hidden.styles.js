import { createScreenMediaQuery } from "../utils/breakpoints";

const getDisplay = (props, size) => (props.hide?.[size] ? "none" : null);

export default (breakpoints) =>
    breakpoints.reduce(
        (styles, breakpoint) => ({
            ...styles,
            [createScreenMediaQuery(breakpoint)]: {
                hidden: {
                    display: (props) => getDisplay(props, breakpoint.size),
                },
            },
        }),
        {
            hidden: {},
        }
    );
