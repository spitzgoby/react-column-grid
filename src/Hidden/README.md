# Hidden

## hide

**Type**:

```Javascript
PropTypes.shape({
    xs: PropTypes.bool,
    sm: PropTypes.bool,
    md: PropTypes.bool,
    lg: PropTypes.bool,
    xl: PropTypes.bool
})
```

**Default**:

```Javascript
{
    xs: false,
    sm: false,
    md: false,
    lg: false,
    xl: false
}
```

**Description**: Determines whether the children will be hidden at a given
breakpoint. If no value is set for a breakpoint then the value from the
next smallest breakpoint will be used.

**Usage**:

```Javascript
// Hides the text when the user's device is xs and shows it for all
// other breakpoints.
<Hidden hide={{ xs: true, sm: false }}>Hide me on xs screens</Hidden>
```

---

## xs

**Type**: `PropTypes.bool`  
**Default**: `false`  
**Description**: Determines whether the children will be hidden at the
extra small breakpoint. This value will only target the `xs` size, and
will not be used at the other breakpoints. If the `hide` property is
also defined it will be given precedence.

---

## sm

**Type**: `PropTypes.bool`  
**Default**: `false`  
**Description**: Determines whether the children will be hidden at the
small breakpoint. This value will only target the `sm` size, and will
not be used at the other breakpoints. If the `hide` property is also
defined it will be given precedence.

---

## md

**Type**: `PropTypes.bool`  
**Default**: `false`  
**Description**: Determines whether the children will be hidden at the
medium breakpoint. This value will only target the `md` size, and
will not be used at the other breakpoints. If the `hide` property is
also defined it will be given precedence.

---

## lg

**Type**: `PropTypes.bool`  
**Default**: `false`  
**Description**: Determines whether the children will be hidden at the
large breakpoint. This value will only target the `lg` size, and
will not be used at the other breakpoints. If the `hide` property is
also defined it will be given precedence.

---

## xl

**Type**: `PropTypes.bool`  
**Default**: `false`  
**Description**: Determines whether the children will be hidden at the
extra large breakpoint. This value will only target the `xl` size, and
will not be used at the other breakpoints. If the `hide` property is
also defined it will be given precedence.
