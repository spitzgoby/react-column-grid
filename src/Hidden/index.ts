import HiddenComponent from "./Hidden";

export const Hidden = HiddenComponent;
export default HiddenComponent;
