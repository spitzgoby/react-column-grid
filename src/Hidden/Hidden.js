import classNames from "classnames";
import { addMissingSizes } from "../utils/breakpoints";
import PropTypes from "prop-types";
import React from "react";

import "./Hidden.scss";

const defaultHide = false;

// Shorthand syntax does not apply to the <Hidden /> element
const useShorthandSyntax = () => false;

const Hidden = ({ children, hide, xs, sm, md, lg, xl }) => {
    const buildHideUsingShorthandSizes = () => ({ xs, sm, md, lg, xl });
    const hasShorthandSizes = xs || sm || md || lg || xl;
    const adjustedHide =
        hide || !hasShorthandSizes
            ? addMissingSizes("hide", hide, defaultHide, useShorthandSyntax)
            : buildHideUsingShorthandSizes();

    const getClass = () =>
        classNames({
            "rcg-h-xs": adjustedHide.xs,
            "rcg-h-sm": adjustedHide.sm,
            "rcg-h-md": adjustedHide.md,
            "rcg-h-lg": adjustedHide.lg,
            "rcg-h-xl": adjustedHide.xl,
        });

    return <span className={getClass()}>{children}</span>;
};

Hidden.propTypes = {
    children: PropTypes.node,
    hide: PropTypes.shape({
        xs: PropTypes.bool,
        sm: PropTypes.bool,
        md: PropTypes.bool,
        lg: PropTypes.bool,
        xl: PropTypes.bool,
    }),
    xs: PropTypes.bool,
    sm: PropTypes.bool,
    md: PropTypes.bool,
    lg: PropTypes.bool,
    xl: PropTypes.bool,
};

export default Hidden;
