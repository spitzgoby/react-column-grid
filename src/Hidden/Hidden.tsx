import classNames from "classnames";
import { GridContext } from "../GridProvider";
import { generateClassNameForSize } from "./Hidden.generateCss";
import PropTypes, { ReactNodeLike } from "prop-types";
import React, { useContext } from "react";
import { addMissingSizes, Breakpoint, sizes } from "../utils/breakpoints";

import { BooleanBreakpointValues } from "../Grid/Grid.layout";

const defaultHide = false;

// Shorthand syntax does not apply to the <Hidden /> element
const useShorthandSyntax = () => false;

type Props = {
    children: ReactNodeLike,
    hide?: BooleanBreakpointValues,
    lg?: boolean,
    md?: boolean,
    sm?: boolean,
    xl?: boolean,
    xs?: boolean
};

const Hidden: React.FC<Props> = ({ children, hide, xs, sm, md, lg, xl }) => {
    const buildHideUsingShorthandSizes = () => ({ xs, sm, md, lg, xl });
    const hasShorthandSizes = xs || sm || md || lg || xl;
    const adjustedHide =
        hide || !hasShorthandSizes
            ? addMissingSizes("hide", hide, defaultHide, useShorthandSyntax)
            : buildHideUsingShorthandSizes();
    const { id } = useContext(GridContext);

    const getClass = () => {
        const classes = sizes.reduce((acc, size: Breakpoint) => {
            acc = {
                ...acc,
                [generateClassNameForSize(size, id)]: adjustedHide[size]
            }

            return acc;
        }, {});

        return classNames(classes);
    }

    return <span className={getClass()}>{children}</span>;
};

Hidden.propTypes = {
    children: PropTypes.node,
    hide: PropTypes.shape({
        xs: PropTypes.bool,
        sm: PropTypes.bool,
        md: PropTypes.bool,
        lg: PropTypes.bool,
        xl: PropTypes.bool,
    }),
    xs: PropTypes.bool,
    sm: PropTypes.bool,
    md: PropTypes.bool,
    lg: PropTypes.bool,
    xl: PropTypes.bool,
};

export default Hidden;
