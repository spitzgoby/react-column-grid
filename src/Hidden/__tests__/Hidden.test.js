import Hidden from "../Hidden";
import React from "react";
import { render } from "@testing-library/react";

describe("<Hidden />", () => {
    it("should render a span component", () => {
        const { container } = render(<Hidden />);

        expect(container).toMatchSnapshot();
    });

    it("should use the correct classes based on size rules", () => {
        const { container } = render(<Hidden xs sm xl />);

        expect(container).toMatchSnapshot();
    });

    it("should prioritize hide prop when present", () => {
        const { container } = render(
            <Hidden hide={{ xs: true, md: false }} md />
        );

        expect(container).toMatchSnapshot();
    });
});
