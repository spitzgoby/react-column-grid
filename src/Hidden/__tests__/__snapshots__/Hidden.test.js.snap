// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Hidden /> should prioritize hide prop when present 1`] = `
<div>
  <span
    class="rcg-h-xs rcg-h-sm"
  />
</div>
`;

exports[`<Hidden /> should render a span component 1`] = `
<div>
  <span
    class=""
  />
</div>
`;

exports[`<Hidden /> should use the correct classes based on size rules 1`] = `
<div>
  <span
    class="rcg-h-xs rcg-h-sm rcg-h-xl"
  />
</div>
`;
